File: LinkedList/LinkedList.cpp
// Linked List Implementation with C++
// Singly linked with basic functionalities only

#include "LinkedList.h"
#include <stdexcept>

// Node
template<class T>
Node<T>::Node(T val, Node* next) {
    this->value = val;
    this->next = next;
}

template<class T>
Node<T>::~Node() {
}

template<class T>
void Node<T>::setValue(T val) {
    this->value = val;
}

template<class T>
void Node<T>::setNext(Node* next) {
    this->next = next;
}

template<class T>
T Node<T>::getValue() {
    return this->value;
}

template<class T>
Node<T>* Node<T>::getNext() {
    return this->next;
}

// Linked List
template<class T>
LinkedList<T>::LinkedList() {
    this->head = nullptr;
}

template<class T>
LinkedList<T>::LinkedList(T const* arr, size_t size) {
    Node<T>* newNode = new Node<T>(arr[0]);
    this->head = newNode;
    Node<T>* ptr = this->head; 
    for (int i = 1; i < size; i++) {
        newNode = new Node<T>(arr[i]);
        ptr->setNext(newNode);
        ptr = newNode;
    }
}

template<class T>
LinkedList<T>::~LinkedList() {
    LinkedList<T>::clear();
}

template<class T>
Node<T>* LinkedList<T>::getHead() {
    return this->head;
}

template<class T>
void LinkedList<T>::setHead(Node<T>* node) {
    this->head = node;
}

template<class T>
T LinkedList<T>::getValue(int pos) {
    Node<T>* ptr = this->head;
    for (int i = 0; i < pos; i++) {
        ptr = ptr->getNext();
        if (ptr == nullptr) {
            throw std::out_of_range("Index Out of Bound");
        }
    }
    return ptr->getValue();
}

template<class T>
void LinkedList<T>::insert(T value, int pos) {
    if (pos == 0) {
        Node<T>* newNode = new Node<T>(value, this->head);
        this->head = newNode;
    } else {
        Node<T>* prevPtr = this->head;
        for (int i = 0; i < pos-1; i++) {
            prevPtr = prevPtr->getNext();
            if (prevPtr->getNext() == nullptr) {
                break;
            }
        }
        Node<T>* newNode = new Node<T>(value, prevPtr->getNext());
        prevPtr->setNext(newNode);
    }
}

template<class T>
void LinkedList<T>::append(T value) {
    Node<T>* newNode = new Node<T>(value);
    Node<T>* ptr = this->head;
    if (ptr == nullptr) {
        this->head = newNode;
        return;
    }
    while (ptr->getNext() != nullptr) {
        ptr = ptr->getNext();
    }
    ptr->setNext(newNode);
}

template<class T>
void LinkedList<T>::remove(int pos) {
    Node<T>* nodeToDelete;
    if (pos == 0) {
        nodeToDelete = this->head;
        this->head = this->head->getNext();
    } else {
        Node<T>* prevNode = this->head;
        for (int i = 0; i < pos-1; i++) {
            prevNode = prevNode->getNext();
            if (prevNode->getNext() == nullptr) {
                throw std::out_of_range("Index Out of Bound");
            }
        }
        nodeToDelete = prevNode->getNext();
        if (nodeToDelete->getNext() == nullptr) {
            prevNode->setNext(nullptr);
        } else {
            prevNode->setNext(nodeToDelete->getNext());
        }
    }
    delete nodeToDelete;
    nodeToDelete = nullptr;
}

template<class T>
void LinkedList<T>::clear() {
    Node<T>* ptr = this->head;
    if (ptr == nullptr) {
        return;
    }
    if (ptr->getNext() == nullptr) {
        delete ptr;
        this->head = nullptr;
        return;
    }
    Node<T>* prevPtr;
    while (ptr != nullptr) {
        prevPtr = ptr;
        ptr = ptr->getNext();
        delete prevPtr;
    }
    this->head = nullptr;
}

template<typename Type>
std::ostream& operator<<(std::ostream &os, LinkedList<Type> &linkedList) {
    os << "[";
    Node<Type>* ptr = linkedList.getHead();
    while(ptr != nullptr) {
        os << ptr->getValue();
        if (ptr->getNext() != nullptr) {
            os << " => ";
        }
        ptr = ptr->getNext();
    }
    os << "]";
    return os;
};

template<typename Type>
LinkedList<Type> operator+(LinkedList<Type> &left, LinkedList<Type> &right) {
    LinkedList<Type> merged;
    Node<Type>* leftPtr = left.getHead();
    Node<Type>* rightPtr = right.getHead();
    while (leftPtr != nullptr && rightPtr != nullptr) {
        if (leftPtr->getValue() < rightPtr->getValue()) {
            merged.append(leftPtr->getValue());
            leftPtr = leftPtr->getNext();
        } else {
            merged.append(rightPtr->getValue());
            rightPtr = rightPtr->getNext();
        }
    }
    while (leftPtr != nullptr) {
        merged.append(leftPtr->getValue());
        leftPtr = leftPtr->getNext();
    }
    while (rightPtr != nullptr) {
        merged.append(rightPtr->getValue());
        rightPtr = rightPtr->getNext();
    }
    return merged;
}

File: LinkedList/LinkedList.py
'''
Linked List Implementation in Python
Singly Linked with basic functionalities only.
'''

class Node:
    __value = None
    __next = None

    def __init__(self, value, next=None):
        self.__value = value
        self.__next = next

    def setValue(self, value):
        self.__value = value

    def setNext(self, next):
        self.__next = next
    
    def getValue(self):
        return self.__value

    def getNext(self):
        return self.__next


class LinkedList:
    __head = None
    
    # constuctor optionally with list object
    def __init__(self, list=None):
        if list is None:
            self.__head = None
        else:
            newNode = Node(list[0])
            self.__head = newNode
            for i in range(1, len(list)):
                ptr = newNode
                newNode = Node(list[i])
                ptr.setNext(newNode)

    
    # get Nth value
    def getVal(self, pos):
        ptr = self.__head
        for i in range(0, pos):
            ptr = ptr.getNext()
            if ptr is None:
                raise IndexError("Index Out of Bound")
        return ptr.getValue()


    # insert at Nth position
    # default to insert to the start if position is not provided
    def insert(self, value, pos=0):
        if pos == 0:
            newNode = Node(value, self.__head)
            self.__head = newNode
        else:
            prevPtr = self.__head
            for i in range(0, pos-1):
                prevPtr = prevPtr.getNext()
                if prevPtr.getNext() is None:
                    break
            newNode = Node(value, prevPtr.getNext())
            prevPtr.setNext(newNode)

    
    # append to the end of list
    def append(self, value):
        newNode = Node(value)
        ptr = self.__head
        if ptr is None:
            self.__head = newNode
            return
        while ptr.getNext() is not None:
            ptr = ptr.getNext()
        ptr.setNext(newNode)


    # remove value at given position
    def remove(self, pos):
        # removing first item
        if pos == 0:
            self.__head = self.__head.getNext()
            return

        prevPtr = self.__head
        for i in range(0, pos-1):
                prevPtr = prevPtr.getNext()
                # if position is out of range
                if prevPtr.getNext() is None:
                    raise IndexError("Index Out of Bound")
        nodeToDelete = prevPtr.getNext()
        # removing last item
        if nodeToDelete.getNext() is None:
            prevPtr.setNext(None)
        else:
            prevPtr.setNext(nodeToDelete.getNext())


    # clear
    def clear(self):
        self.__head = None

    
    # iterator
    def iterator(self):
        return self.__head


    # print format
    def __str__(self):
        toString = "{"
        ptr = self.__head
        while ptr != None:
            toString += str(ptr.getValue())
            if ptr.getNext() != None:
                toString += " => "
            ptr = ptr.getNext()
        toString += "}"
        return toString

    
    # add two sorted lists in a new sorted list
    # this is used for Merge Two Sorted List specifically
    # so assume two linked lists are already sorted.
    def __add__(self, other):
        comb = LinkedList()
        selfPtr = self.__head
        otherPtr = other.__head
        while selfPtr is not None and otherPtr is not None:
            if selfPtr.getValue() < otherPtr.getValue():
                comb.append(selfPtr.getValue())
                selfPtr = selfPtr.getNext()
            else:
                comb.append(otherPtr.getValue())
                otherPtr = otherPtr.getNext()
        while selfPtr is not None:
            comb.append(selfPtr.getValue())
            selfPtr = selfPtr.getNext()
        while otherPtr is not None:
            comb.append(otherPtr.getValue())
            otherPtr = otherPtr.getNext()
        return comb
